src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -74,44 +74,123 @@ public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAl
+                    Log.FailedWritingMessage(_logger, ex);
+                }
+            }
+
+            // Otherwise, release the lock
+            _writeLock.Release();
+
+            return Task.CompletedTask;
+        }
 
+        private ValueTask<FlushResult> WriteCore(HubMessage message)
+        {
             try
             {
-                // This will internally cache the buffer for each unique HubProtocol/DataEncoder combination
+                // This will internally cache the buffer for each unique HubProtocol
                 // So that we don't serialize the HubMessage for every single connection
                 var buffer = message.WriteMessage(Protocol);
+
                 _connectionContext.Transport.Output.Write(buffer);
 
-                Interlocked.Exchange(ref _lastSendTimestamp, Stopwatch.GetTimestamp());
+                return _connectionContext.Transport.Output.FlushAsync();
+            }
+            catch (Exception ex)
+            {
+                Log.FailedWritingMessage(_logger, ex);
 
-                await _connectionContext.Transport.Output.FlushAsync();
+                return new ValueTask<FlushResult>(new FlushResult(isCanceled: false, isCompleted: true));
+            }
+        }
+
+        private async Task CompleteWriteAsync(ValueTask<FlushResult> task)
+        {
+            try
+            {
+                await task;
+            }
+            catch (Exception ex)
+            {
+                Log.FailedWritingMessage(_logger, ex);
+            }
+            finally
+            {
+                _writeLock.Release();
+            }
+        }
+
+        private async Task WriteSlowAsync(HubMessage message)
+        {
+            try
+            {
+                await _writeLock.WaitAsync();
+
+                await WriteCore(message);
+            }
+            catch (Exception ex)
+            {
+                Log.FailedWritingMessage(_logger, ex);
             }
             finally
             {
                 _writeLock.Release();
             }
         }
 
-        private async Task TryWritePingAsync()
+        private Task TryWritePingAsync()
         {
             // Don't wait for the lock, if it returns false that means someone wrote to the connection
             // and we don't need to send a ping anymore
-            if (!await _writeLock.WaitAsync(0))
+            if (!_writeLock.Wait(0))
             {
-                return;
+                return Task.CompletedTask;
             }
 
+            return TryWritePingSlowAsync();
+        }
+
+        private async Task TryWritePingSlowAsync()
+        {
             try
             {
                 Debug.Assert(_cachedPingMessage != null);
-                _connectionContext.Transport.Output.Write(_cachedPingMessage);
 
-                Interlocked.Exchange(ref _lastSendTimestamp, Stopwatch.GetTimestamp());
+                _connectionContext.Transport.Output.Write(_cachedPingMessage);
```

Comment:
Yeah, in reviewing the WebSockets code, this is basically how it's pings work. It sends them at the ping interval every time, regardless of if another message was sent. But it won't wait on the send lock.

A side effect of this, is that Long Poll requests will **always** complete in at least `KeepAliveInterval`. We could lower the timeout on the client to something nearer to this value in theory.

Reply:
Never mind, I forgot my own code. Keep-Alives aren't sent for long polling because of my favo**u**rite feature: `IConnectionInherentKeepAliveFeature`